### PR TITLE
WLT-424: heavy pulse completion (HME side)

### PR DESCRIPTION
**Pull request by uhzxl**
**- What I did**
Added pulse completion on HME
**- How I did it**
Added jetKeeper, db jet.Storage and pulseValidator to eventually get a complete pulse on HME
**- How to verify it**
Run `./scripts/insolard/launchnet.sh -g` and find sequence of records `[PulseValidator] pulse #%v complete` in HME logs

### CODE REVIEW

File: ledger/heavy/consistency/pulse.go [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+	"github.com/insolar/insolar/insolar"
+	"github.com/insolar/insolar/insolar/jet"
+	"github.com/insolar/insolar/insolar/utils"
+	"github.com/insolar/insolar/instrumentation/inslogger"
+	"github.com/insolar/insolar/internal/ledger/store"
+	"github.com/insolar/insolar/ledger/heavy/replica"
+)
+
+type PulseValidator interface {
```
Comment (andreyromancev) [inline ledger/heavy/consistency/pulse.go]: Discussed: moving to JetKeeper.

File: ledger/heavy/handler/handler.go [hunk 2]
```diff
@@ -351,20 +359,48 @@ func (h *Handler) getCode(ctx context.Context, id *insolar.ID) (record.Code, err
...
-		return &reply.HeavyError{Message: err.Error(), JetID: msg.JetID, PulseNum: msg.PulseNum}, nil
+
+	{
+		pulse, err := h.PulseAccessor.Latest(ctx)
+		if err != nil {
+			return &reply.HeavyError{Message: err.Error(), JetID: msg.JetID, PulseNum: msg.PulseNum}, nil
+		}
+		futurePulse := pulse.NextPulseNumber
+		drop, err := storeDrop(ctx, h.DropModifier, msg.Drop)
+		switch {
```
Comment (andreyromancev) [inline ledger/heavy/handler/handler.go]: First two cases do `return` at the end., so `switch` is not necessary. We can flatten it.

File: ledger/heavy/pulsemanager/pulsemanager.go [hunk 3]
```diff
@@ -94,12 +98,26 @@ func (m *PulseManager) Set(ctx context.Context, newPulse insolar.Pulse, persist
...
+		oldPulse *insolar.Pulse
+	)
+
 	m.GIL.Acquire(ctx)
 	ctx, span := instracer.StartSpan(ctx, "pulse.gil_locked")
 
 	defer span.End()
 	defer m.GIL.Release(ctx)
 
+	// FIXME: @andreyromancev. 17.12.18. return insolar.Pulse here.
```
Comment (andreyromancev) [inline ledger/heavy/pulsemanager/pulsemanager.go]: Let's remove this `FIXME`.

File: ledger/heavy/replica/jetkeeper.go [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+	"github.com/pkg/errors"
+
+	"github.com/insolar/insolar/insolar"
+	"github.com/insolar/insolar/insolar/utils"
+	"github.com/insolar/insolar/internal/ledger/store"
+)
+
+type JetKeeper interface {
+	Add(insolar.PulseNumber, insolar.JetID) error
+	All(insolar.PulseNumber) ([]insolar.JetID, error)
```
Comment (andreyromancev) [inline ledger/heavy/replica/jetkeeper.go]: This method is not necessary if we update synced jets inside `Add`.

File: insolar/jet.go [hunk 5]
```diff
@@ -122,6 +122,21 @@ ScanPrefix:
 	return res.String()
 }
 
+type JetIDSlice []JetID
```
[the remark this answers is not in the document]
Comment (uhzxl) [reply insolar/jet.go]: done

File: insolar/jet/dbstore.go [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+
+func (s *DBStore) ForID(ctx context.Context, pulse insolar.PulseNumber, recordID insolar.ID) (insolar.JetID, bool) {
+	s.RLock()
+	defer s.RUnlock()
+
+	tree := s.get(pulse)
+	return tree.Find(recordID)
+}
+
+func (s *DBStore) Update(ctx context.Context, pulse insolar.PulseNumber, actual bool, ids ...insolar.JetID) {
```
[the remark this answers is not in the document]
Comment (uhzxl) [reply insolar/jet/dbstore.go]: done

File: insolar/jet/dbstore.go [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+	if err != nil {
+		return insolar.ZeroJetID, insolar.ZeroJetID, err
+	}
+	err = s.set(pulse, tree)
+	if err != nil {
+		return insolar.ZeroJetID, insolar.ZeroJetID, err
+	}
+	return left, right, nil
+}
+func (s *DBStore) Clone(ctx context.Context, from, to insolar.PulseNumber) {
```
[the remark this answers is not in the document]
Comment (uhzxl) [reply insolar/jet/dbstore.go]: done

File: insolar/jet/dbstore.go [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+
+	tree := s.get(from)
+	newTree := tree.Clone(false)
+	err := s.set(to, newTree)
+	if err != nil {
+		logger.Error(errors.Wrapf(err, "failed to clone jet.Tree"))
+	}
+}
+
+func (s *DBStore) DeleteForPN(ctx context.Context, pulse insolar.PulseNumber) {
```
[the remark this answers is not in the document]
Comment (uhzxl) [reply insolar/jet/dbstore.go]: done

File: insolar/jet/dbstore.go [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+}
+
+type pulseKey insolar.PulseNumber
+
+func (k pulseKey) Scope() store.Scope {
+	return store.ScopeJetTree
+}
+
+func (k pulseKey) ID() []byte {
+	return utils.UInt32ToBytes(uint32(k))
```
[the remark this answers is not in the document]
Comment (uhzxl) [reply insolar/jet/dbstore.go]: done

File: insolar/jet/dbstore_test.go [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+
+func TestDBStorage_ForID_Basic(t *testing.T) {
+	ctx := inslogger.TestContext(t)
+
+	pn := gen.PulseNumber()
+	meaningfulBits := "01000011" + "11000011" + "010010"
+
+	bits := parsePrefix(meaningfulBits)
+	expectJetID := NewIDFromString(meaningfulBits)
+	// fmt.Printf("expectJetID:        %08b\n", expectJetID[:])
```
[the remark this answers is not in the document]
Comment (uhzxl) [reply insolar/jet/dbstore_test.go]: cleaned up

File: ledger/heavy/consistency/pulse.go (hunk 1 above)
Comment (uhzxl) [reply ledger/heavy/consistency/pulse.go]: done

File: ledger/heavy/handler/handler.go (hunk 2 above)
Comment (uhzxl) [reply ledger/heavy/handler/handler.go]: done

File: ledger/heavy/pulsemanager/pulsemanager.go (hunk 3 above)
Comment (uhzxl) [reply ledger/heavy/pulsemanager/pulsemanager.go]: done

File: ledger/heavy/replica/jetkeeper.go (hunk 4 above)
Comment (uhzxl) [reply ledger/heavy/replica/jetkeeper.go]: done

File: insolar/jet/dbstore_test.go [hunk 11]
```diff
@@ -0,0 +1,236 @@
...
+	s.Clone(ctx, 100, 101)
+
+	tree = dbTreeForPulse(s, 101)
+	require.Equal(t, "root (level=0 actual=false)\n", tree.String())
+
+	tree = dbTreeForPulse(s, 100)
+	require.Equal(t, "root (level=0 actual=true)\n", tree.String())
+}
+
+// func TestDBStorage_DeleteJetTree(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (uhzxl) [reply insolar/jet/dbstore_test.go]: done

File: insolar/jet/dbstore_test.go [hunk 12]
```diff
@@ -0,0 +1,236 @@
...
+func TestDBStorage_UpdateJetTree(t *testing.T) {
+	ctx := inslogger.TestContext(t)
+
+	db := store.NewMemoryMockDB()
+	s := NewDBStore(db)
+
+	s.Update(ctx, 100, true, *insolar.NewJetID(0, nil))
+
+	tree := dbTreeForPulse(s, 100)
+	require.Equal(t, "root (level=0 actual=true)\n", tree.String())
```
[the remark this answers is not in the document]
Comment (uhzxl) [reply insolar/jet/dbstore_test.go]: fixed

File: insolar/jet/dbstore_test.go [hunk 13]
```diff
@@ -0,0 +1,236 @@
...
+func TestDBStorage_CloneJetTree(t *testing.T) {
+	ctx := inslogger.TestContext(t)
+
+	db := store.NewMemoryMockDB()
+	s := NewDBStore(db)
+
+	s.Update(ctx, 100, true, *insolar.NewJetID(0, nil))
+
+	tree := dbTreeForPulse(s, 100)
+	require.Equal(t, "root (level=0 actual=true)\n", tree.String())
```
[the remark this answers is not in the document]
Comment (uhzxl) [reply insolar/jet/dbstore_test.go]: fixed

File: insolar/jet/dbstore_test.go [hunk 14]
```diff
@@ -0,0 +1,236 @@
...
+func TestDBStorage_SplitJetTree(t *testing.T) {
+	ctx := inslogger.TestContext(t)
+
+	db := store.NewMemoryMockDB()
+	s := NewDBStore(db)
+
+	root := insolar.NewJetID(0, nil)
+	left, right, err := s.Split(ctx, 100, *root)
+	require.NoError(t, err)
+	require.Equal(t, "[JET 0 -]", root.DebugString())
```
[the remark this answers is not in the document]
Comment (uhzxl) [reply insolar/jet/dbstore_test.go]: fixed

**Pull request merged**